src/main/java/co/cask/gcp/bigquery/BigQuerySink.java

```diff
@@ -244,6 +248,12 @@ private void emitLineage(BatchSinkContext context, List<BigQueryTableFieldSchema
   }
 
   private static void decodeSimpleTypes(JsonObject json, String name, StructuredRecord input) {
+    // Avoid NPE when structured record schema does not contain field present in output schema.
+    if (input.getSchema().getField(name) == null) {
```

Comment:
this can be avoided if the method is changed so that instead of taking `String name, StructuredRecord input`, it takes `Schema.Field field, Object fieldVal`. The caller should not iterate through config.getSchema().getFields(), but through input.getSchema().getFields(). If the record field does not exist in the config schema, we can skip it. Otherwise it should call decodeSimpleTypes()

Reply:
also, the original bug was if the config schema contained extra non-nullable fields that do not exist in the input schema. Checking here will cause it to fail late, at runtime.

This can be checked at configure time if there are no macros, using pipelineConfigurer().getStageConfigurer().getInputSchema() to get the input schema if it's known. If the configured schema contains a macro, it can be checked against the input schema at prepare time using context.getInputSchema().

Reply:
[1.] Modified the code to iterate over record schema
[2.] Added checks in schema validation to fail fast.